### PR TITLE
fix(kv_index): purge tree tenants when a worker is removed

### PR DESCRIPTION
**Pull request by slin1237**
### Problem

`CacheAwarePolicy::remove_worker` / `remove_worker_by_url` are no-op stubs that rely on LRU eviction to reclaim a removed worker's radix-tree state. But `evict_tenant_by_size` only evicts tenants whose per-tenant count exceeds `max_tree_size`, and a removed worker's counter never grows again — so its subtree (up to the budget) is retained forever, its tenant entries keep shared nodes non-empty (pinning other tenants' evicted paths), and node splits keep copying the dead tenant into new intermediate nodes. On a large fleet with heavy pod churn this is a steady router RSS leak.

### Solution

Purge the worker's tenant eagerly when it is removed:

- The token tree gains `remove_tenant_all`, mirroring the string tree's (previously caller-less) API: remove the tenant from every node including the root, detach nodes the removal emptied via the existing eviction ancestor cleanup, and drop the `tenant_token_count` entry, debiting the tree-wide token total by the removed count.
- The string tree's `remove_tenant_all` now also detaches nodes the removal emptied (same empty-node rule as `evict_tenant_by_size`) instead of leaving unreachable empty nodes behind, and keeps dropping the `tenant_char_count` entry (which also debits the tree-wide char total).
- `remove_worker` / `remove_worker_by_url` purge the worker's URL from every model's string and token tree. `hash_index` needs no per-worker purge: its entries are keyed by node-path hash and carry no tenant.

## Changes

- `crates/kv_index/src/token_tree.rs`: add `TokenTree::remove_tenant_all` + `collect_tenant_nodes`, reusing `remove_tenant_and_cleanup` for empty-node detachment
- `crates/kv_index/src/string_tree.rs`: `remove_tenant_all` detaches emptied nodes via new `detach_empty_ancestors` (mirrors the eviction empty-node rule)
- `model_gateway/src/policies/cache_aware.rs`: implement `remove_worker` / `remove_worker_by_url` (previously no-op stubs)
- Tests: per-tree purge tests (count entry gone, surviving tenant still matches its full prefix, sole-owner branches detached) and a policy-level add/route/remove test asserting tenant counts empty for the removed URL in both trees; the token tree's total-vs-tenant-sum consistency test gains a tenant-purge step (matching the string tree's)

## Test Plan

- `cargo test -p kv-index` — 220 passed, 0 failed (includes 4 new `remove_tenant_all` tests covering both trees)
- `cargo test -p smg --lib policies` — 132 passed, 0 failed (includes new `test_remove_worker_purges_tree_tenants`)
- `cargo test -p smg` — all test binaries pass
- `cargo clippy --all-targets -- -D warnings` — clean
- `cargo +nightly fmt --all` — clean

<details>
<summary>Checklist</summary>

- [x] `cargo +nightly fmt` passes
- [x] `cargo clippy --all-targets --all-features -- -D warnings` passes
- [ ] (Optional) Documentation updated
- [ ] (Optional) Please join us on Slack [#sig-smg](https://slack.lightseek.org) to discuss, review, and merge PRs

</details>

### CODE REVIEW

File: crates/kv_index/src/string_tree.rs [hunk 1]
```diff
@@ -1116,10 +1113,28 @@ impl Tree {
...
+        let mut current = Arc::clone(node);
+        while current.children.is_empty() && current.tenant_last_access_time.is_empty() {
+            let Some(parent) = current.parent.read().as_ref().and_then(Weak::upgrade) else {
+                break;
+            };
+            if let Some(fc) = current.text.read().first_char() {
+                parent.children.remove(&fc);
+            }
+            current = parent;
+        }
```
[the remark this answers is not in the document]
Comment (slin1237) [reply crates/kv_index/src/string_tree.rs:1201]: The check-then-detach window described here is the tree's established concurrency model, not something this PR introduces — both purge paths reuse the existing eviction primitives:

- Token tree: `remove_tenant_all` calls the same `remove_tenant_and_cleanup` / `cleanup_empty_ancestors_with_parent` that `evict_tenant` has always used; the unconditional `parent.children.remove(&page_key)` predates this PR.
- String tree: `detach_empty_ancestors` reproduces the `evict_tenant_by_size` empty-node rule (string_tree.rs L1017-L1024), which does the same unconditional `parent.children.remove(&fc)`.

That window is already exercised concurrently today: `CacheAwarePolicy` runs `evict_tenant_by_size` on a background `PeriodicTask` every `eviction_interval_secs` against live trees while request hot paths insert, with no structural lock. The trees are lock-free by design and documented as eventually consistent under concurrent structural mutation (see `TokenTree::iter_entries`).

The impact is also narrower than "removes routes for other tenants": these trees are advisory affinity indexes, not routing state. If a purge detaches a just-split or just-repopulated branch, the effect is a lost cache-affinity hint for that prefix — selection falls back to min-load/first-healthy (`select_worker_*` never requires tree completeness) and the next insert re-creates the path. Tenant counts use saturating arithmetic.

An identity-checked removal (`remove_if` + `Arc::ptr_eq`) would not close the race either: an in-flight descent already holding an `Arc` to the node can repopulate it after the emptiness re-check. A real guarantee requires serializing insert/split/purge behind a structural lock on the request hot path, across both trees plus the existing eviction path — a tree-wide concurrency redesign out of scope for this leak fix. Deterministic regression coverage for the window has the same dependency (it needs injectable sync points; a stress loop would be flaky by construction).

Keeping the purge on the same primitives and semantics as the shipped eviction is intentional.
